### PR TITLE
Issue 2183

### LINKED ISSUE #2183: Enable reordering symbolizers
# Feature Request

**Is your feature request related to a problem? Please describe.**
Order of symbolizers affects the resulting symbol. Eg. outline color of fill symbolizer can completely cover polyline if line symbolizer is defined after the fill.

**Describe the solution you'd like**
Enable to change order of symbolizers, ideally with drag&drop. It would be also nice to be able to collapse/expand each symbolizer panel.

### PR DESCRIPTION
**Pull request by DailisLangovskis**
## Description

Introduces drag and drop feature for styler rules and symbolizers

## Related issues or pull requests

closes #2183 #2150 
refs #2148 

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [x] Feature
- [ ] Dependency updates
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe)
- [ ] I am unsure (we'll look into it together)

## Do you introduce a breaking change?

- [ ] Yes
- [x] No
- [ ] I am unsure (no worries, we'll find out)

## Checklist

- [x] I understand and agree that the changes in this PR will be licensed under the [MIT License]
- [x] I have followed the [guidelines for contributing](https://github.com/hslayers/hslayers-ng/blob/master/CONTRIBUTING.md)
- [x] The proposed change fits to the content of the [code of conduct](https://github.com/hslayers/hslayers-ng/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have added or updated tests and documentation, and the test suite passes (run `npm test` locally)
- [ ] I'm lost; why do I have to check so many boxes? Please help!

### CODE REVIEW

**Comment by DailisLangovskis** [comment]
@FilipLeitner I would appreciate if You could add the missing translations as well as, if it is possible to fix the symbolizer type dropdown, that is getting opened behind other DOM elements, couldn't get it right.

File: projects/hslayers/src/components/styles/styles.module.ts [hunk 1]
```diff
@@ -1,10 +1,13 @@
+import {BrowserAnimationsModule} from '@angular/platform-browser/animations';
 import {CUSTOM_ELEMENTS_SCHEMA, NgModule} from '@angular/core';
 import {CommonModule} from '@angular/common';
+import {DragDropModule} from '@angular/cdk/drag-drop';
 import {FormsModule} from '@angular/forms';
-import {TranslateModule} from '@ngx-translate/core';
+import {MatExpansionModule} from '@angular/material/expansion';
```
Comment (raitisbe) [inline projects/hslayers/src/components/styles/styles.module.ts:6]: I dont like to mix material design into our bootstrap version of hslayers. Becomes too much of a Frankenstein.

Comment (DailisLangovskis) [reply projects/hslayers/src/components/styles/styles.module.ts:6]: I understand, but couldn't get drag and drop to work on ngb

Comment (DailisLangovskis) [reply projects/hslayers/src/components/styles/styles.module.ts:6]: I will try to make it work with ngb then.

Comment (DailisLangovskis) [reply projects/hslayers/src/components/styles/styles.module.ts:6]: How ever we could consider to move to angular-material at some point and remove ngb dependencies. (Just a thought)

Comment (raitisbe) [reply projects/hslayers/src/components/styles/styles.module.ts:6]: I think the correct way would be to have a UI kit agnostic hslayers core and then have 2 more extensions/libraries - one for NgBootstrap one for Material

**Comment by FilipLeitner** [comment]
> @FilipLeitner I would appreciate if You could add the missing translations as well as check the css and give me a notice on some errors, feel free to give me some suggestions. Also when adding the styler.component.scss file as import to global app.scss, it causes some build error, coming from overwrite functions.

Sorry I must have missed the @ mention email notification among all those dependency upgrade ones so I didnt even know you that u asked for anything. Will do ! :)

**Comment by raitisbe** [comment]
The correct conventiopnal commits prefix is `feat` instead of `feature` ;)

**Comment by FilipLeitner** [comment]
btw :D I spent some quality time trying to get that styling to work some other way  but because of the nested cdk dnd it might as well stay as it is.
- compilation error is bacuase of unexpected value being passed to to-hsl function (rgba definition not strarting with 0) . Could be solved by coverting rgba to hsla. If thats done and styler.component.scss declarations are removed from styleUrls we need to style dnd preview ourselves. Its a bit tricky as it gets inserted inside body and styles not getting passed to children + more.

...so long story short Id leave it as it is (Y) . ill try to provide the  translations at least :P

**Comment by FilipLeitner** [comment]
Wonder where could the 'Reordering' toggle button go. Its not really convenient up there in my opinion. Had to go through code to figure it out :D

**Comment by raitisbe** [comment]
I think we don't need the button at all and reordering can be toggled on by default.

**Comment by DailisLangovskis** [comment]
> The correct conventional commits prefix is `feat` instead of `feature` ;)

Sorry, my bad. Will reword it

**Review by raitisbe — APPROVED** [review]

**Review by FilipLeitner — APPROVED** [review]

**Comment by jmacura** [comment]
The drag'n'drop feature is really cool, but a bit eager. It starts dragging when I click&move the mouse button _anywhere_ in the styler. It means, when I want to play with the range buttons like Opacity or Width, it starts dragging instead...
image

**Comment by DailisLangovskis** [comment]
> The drag'n'drop feature is really cool, but a bit eager. It starts dragging when I click&move the mouse button _anywhere_ in the styler. It means, when I want to play with the range buttons like Opacity or Width, it starts dragging instead...
> image

Well that is why I had this toogle for dragging, to avoid it, that I removed on the request. Must admin, didn't notice it when using sliders.

**Comment by jmacura** [comment]
This PR does not fully resolve #2150 . Markers are not reflected in legend.
image

**Comment by jmacura** [comment]
> 
> Well that is why I had this toogle for dragging, to avoid it, that I removed on the request. Must admin, didn't notice it when using sliders.

Is it not possible to make the block draggable only by its header?
This I mean: 
image

**Comment by DailisLangovskis** [comment]
> > Well that is why I had this toogle for dragging, to avoid it, that I removed on the request. Must admin, didn't notice it when using sliders.
> 
> Is it not possible to make the block draggable only by its header?
> This I mean:
> image

Well the thing is that tha cdkDrag item is a component, not a one div or smth. Had to create a component list item, because there are no uniq identifiers inside styleOnject rules or symbolizers I could relay on the check which rule was opened and whick symbolizer.

**Comment by DailisLangovskis** [comment]
> > > Well that is why I had this toogle for dragging, to avoid it, that I removed on the request. Must admin, didn't notice it when using sliders.
> > 
> > 
> > Is it not possible to make the block draggable only by its header?
> > This I mean:
> > image
> 
> Well the thing is that tha cdkDrag item is a component, not a one div or smth. Had to create a component list item, because there are no uniq identifiers inside styleOnject rules or symbolizers I could relay on the check which rule was opened and whick symbolizer.

In this case the answer would be no.

**Comment by raitisbe** [comment]
> > > > Well that is why I had this toogle for dragging, to avoid it, that I removed on the request. Must admin, didn't notice it when using sliders.
> > > 
> > > 
> > > Is it not possible to make the block draggable only by its header?
> > > This I mean:
> > > image
> > 
> > 
> > Well the thing is that the cdkDrag item is a component, not a one div or smth. Had to create a component list item, because there are no uniq identifiers inside styleObject rules or symbolizers I could relay on to check,which rule was opened and which symbolizer.
> 
> In this case the answer would be no.

Could search a bit about event bubbling and clickthrough, maybe something comes up. It is possible to handle mouse events only in nested components, but not on the parent.

**Comment by raitisbe** [comment]
Also: https://stackoverflow.com/questions/57710285/set-a-part-of-a-draggable-div-as-undraggable

**Comment by DailisLangovskis** [comment]
> This PR does not fully resolve #2150 .
> 
> Markers are not reflected in legend.
> image
> 
> Neither the Fill-border vs. Line clash is addressed.
> image

The clash gets resolved by draging symbolizer order, the only way to do it.

**Comment by DailisLangovskis** [comment]
> This PR does not fully resolve #2150 .
> 
> Markers are not reflected in legend.
> image

This must be addressed in a seperate PR, because it is not an easy small fix to implement!

**Comment by DailisLangovskis** [comment]
> The drag'n'drop feature is really cool, but a bit eager. It starts dragging when I click&move the mouse button _anywhere_ in the styler. It means, when I want to play with the range buttons like Opacity or Width, it starts dragging instead...
> image

Resolved!

**Comment by fzadrazil** [comment]
Reordering of symbolizers works great now, thank you. Just one more minor request. When I add a new symbolizer, the dropdown list with symbolizer types stays open now. It should close automatically. Testing this in the hslayers/test-app project.

**Review by fzadrazil — APPROVED** [review]

**Pull request merged**